src/main/java/gregtech/common/inventory/appeng/SerializableFluidList.java

```diff
@@ -0,0 +1,144 @@
+package gregtech.common.inventory.appeng;
+
+import appeng.api.config.FuzzyMode;
+import appeng.api.storage.data.IAEFluidStack;
+import appeng.api.storage.data.IItemList;
+import appeng.fluids.util.AEFluidStack;
+import appeng.fluids.util.MeaningfulFluidIterator;
+import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
+import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap;
+import net.minecraft.nbt.NBTBase;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraftforge.common.util.INBTSerializable;
+import net.minecraftforge.fluids.Fluid;
+import net.minecraftforge.fluids.FluidStack;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+
+/**
+ * @Author GlodBlock
+ * @Description A serializable {@link IItemList} from AE2, specially improved for fluid.
+ * @Date 2023/4/18-23:52
+ */
+public class SerializableFluidList implements IItemList<IAEFluidStack>, INBTSerializable<NBTTagList> {
+
+    private final Reference2ObjectMap<Fluid, IAEFluidStack> records = new Reference2ObjectOpenHashMap<>();
+
+    public SerializableFluidList() {
+    }
+
+    @Override
+    public void add(IAEFluidStack fluid) {
+        if (fluid != null) {
+            this.getOrCreateRecord(fluid).add(fluid);
+        }
+    }
+
+    @Override
+    public IAEFluidStack findPrecise(IAEFluidStack fluid) {
+        if (fluid == null) {
+            return null;
+        } else {
+            return this.records.get(fluid.getFluid());
+        }
+    }
+
+    @Override
+    public Collection<IAEFluidStack> findFuzzy(IAEFluidStack filter, FuzzyMode fuzzy) {
+        return Collections.singleton(findPrecise(filter));
```

Comment:
should return a empty collection if the filter is null, like the FluidList from AE2

Reply:
done